proof/crefine/RISCV64/Syscall_C.thy

```diff
@@ -94,9 +94,592 @@ lemma performInvocation_Notification_ccorres:
+   apply ccorres_remove_UNIV_guard
+   apply (rule ccorres_getSlotCap_cte_at)
+   apply (rule ccorres_move_array_assertion_tcb_ctes)
+   apply (rule ccorres_split_nothrow)
+       apply (ctac add: getSlotCap_h_val_ccorres)
+      apply ceqv
+     apply (rule ccorres_move_c_guard_tcb)
+     apply (rule ccorres_pre_threadGet_P)
+     apply (ctac add: sendFaultIPC_ccorres)
+       apply (simp add: unless_def when_def)
+       apply (rule ccorres_cond_both'[where Q=\<top> and Q'=\<top>])
+         apply (clarsimp simp: to_bool_def)
+        apply (ctac add: handleNoFaultHandler_ccorres)
+       apply (rule ccorres_return_Skip)
+      apply ((wpsimp wp: sendFaultIPC_invs' hoare_drop_imps | strengthen invs'_implies)+)[1]
+     apply (vcg exspec=sendFaultIPC_modifies)
+    apply (wpsimp wp: hoare_vcg_all_lift hoare_drop_imps getSlotCap_wp)
+   apply vcg
+  apply clarsimp
+  apply (frule invs_iflive')
+  apply (frule (1) st_tcb_ex_cap'')
+   apply fastforce
+  apply (frule st_tcb_strg'[rule_format])
+  apply normalise_obj_at'
+  apply (frule cap_in_tcbFaultHandlerSlot)
+   apply fastforce
+  apply (clarsimp simp: tcb_cnode_index_defs tcbSlots cte_level_bits_def)
+  apply (intro conjI impI allI)
+   apply fastforce
+  apply (rename_tac tcb)
+  apply clarsimp
+  apply (frule (1) obj_at_cslift_tcb)
+  apply normalise_obj_at'
+  apply (frule_tac k=tcb in tcb_ko_at_valid_objs_valid_tcb')
+   apply fastforce
+  apply (clarsimp simp: from_bool_def typ_heap_simps valid_tcb'_def ctcb_relation_def
+                 split: if_splits)
+  apply (case_tac "tcbSchedContext tcb"; clarsimp)
+  done
+
+lemma validTimeoutHandler_ccorres:
+  "ccorres (\<lambda>rv rv'. rv = to_bool rv') ret__unsigned_long_'
+     (tcb_at' tptr) \<lbrace>\<acute>tptr = tcb_ptr_to_ctcb_ptr tptr\<rbrace> hs
+     (isValidTimeoutHandler tptr) (Call validTimeoutHandler_'proc)"
+  apply (cinit lift: tptr_')
+   apply (rule ccorres_stateAssert)
+   apply (simp add: getThreadTimeoutHandlerSlot_def locateSlot_conv cte_C_size)
+   apply ccorres_remove_UNIV_guard
+   apply (rule ccorres_getSlotCap_cte_at)
+   apply (rule ccorres_move_array_assertion_tcb_ctes)
+   apply (rule ccorres_split_nothrow)
+       apply (ctac add: getSlotCap_h_val_ccorres)
+      apply ceqv
+     apply csymbr
+     apply (wpc; fastforce intro: ccorres_return_C simp: cap_get_tag_isCap isCap_simps)
+    apply wpsimp
+   apply vcg
+  apply (fastforce simp: tcb_cnode_index_defs tcbSlots cte_level_bits_def)
+  done
+
+crunch doIPCTransfer, handleFaultReply
+  for weak_sch_act_wf[wp]: "\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s"
+  and pspace_aligned'[wp]: pspace_aligned'
+  and pspace_distinct'[wp]: pspace_distinct'
+  and pspace_bounded'[wp]: pspace_bounded'
+  and no_0_obj'[wp]: no_0_obj'
+  (wp: crunch_wps)
+
+lemma tcbFault_update_valid_objs'[wp]:
+  "threadSet (tcbFault_update f) tcbPtr \<lbrace>valid_objs'\<rbrace>"
+  by (wpsimp wp: threadSet_valid_objs')
+
+lemma doReplyTransfer_ccorres:
```

Comment:
is this proof largely derived from an existing one, or mostly new?

Reply:
I can't remember all that well, but I'm pretty sure it's almost all new. Maybe a few bits copy pasted from the old proof.